File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/lang/ChineseLanguagePack.kt
package com.draftright.keyboard.lang

import com.draftright.keyboard.Composer
import com.draftright.keyboard.KeyDef
import com.draftright.keyboard.LanguagePack
import com.draftright.keyboard.composer.PinyinComposer
import com.draftright.keyboard.ime.CandidateEngine
import com.draftright.keyboard.ime.ChinesePinyinSeedDictionary
import com.draftright.keyboard.ime.DictPackResolver
import com.draftright.keyboard.ime.DictionaryCandidateEngine
import com.draftright.keyboard.ime.ImeContext
import java.util.Locale

/**
 * Chinese (中文) pinyin pack — type pinyin on the standard QWERTY ([QwertyLayout]);
 * [PinyinComposer] shows the pinyin live and [DictionaryCandidateEngine] offers
 * Hanzi candidates. Same shape as Japanese, only the dictionary differs (Rule #1).
 *
 * Ships a built-in seed; the full pinyin→hanzi dictionary arrives later as a
 * downloadable pack (mirrors the Japanese dict-pack pipeline).
 */
object ChineseLanguagePack : LanguagePack {
    override val id = "zh"
    override val displayName = "中文"
    override val locale: Locale = Locale.CHINESE

    override val alphaRows: List<List<KeyDef>> = QwertyLayout.alphaRows
    override val symbols1Rows: List<List<KeyDef>> = QwertyLayout.symbols1Rows
    override val symbols2Rows: List<List<KeyDef>> = QwertyLayout.symbols2Rows
    override val longPressAccents: Map<Char, List<Char>> = emptyMap()

    override fun composer(): Composer = PinyinComposer()

    /** Matches the backend manifest's pack URL prefix (zh pack ships later). */
    private const val PACK_ID_PREFIX = "draftright-ime-zh"

    /**
     * Cached engine + the pack identity it was built from. Parsing the pack is
     * expensive, so keep the engine across keystrokes — but key the cache on the
     * resolved pack file (its name encodes the version) so a pack installed
     * mid-session rebuilds instead of serving the stale seed/dict (issue #10).
     * Per-keystroke cost is one small directory scan.
     */
    @Volatile
    private var cachedEngine: CandidateEngine? = null
    @Volatile
    private var cachedKey: String? = null

    override fun candidateEngine(): CandidateEngine {
        val ctx = ImeContext.appOrNull()
        val key = if (ctx != null) {
            DictPackResolver.resolvedPackFile(ctx, PACK_ID_PREFIX)?.path ?: "seed"
        } else {
            "seed"
        }
        cachedEngine?.let { if (key == cachedKey) return it }
        synchronized(this) {
            cachedEngine?.let { if (key == cachedKey) return it }
            val dict = if (ctx != null) {
                DictPackResolver.loadOrFallback(ctx, PACK_ID_PREFIX) { ChinesePinyinSeedDictionary.dict }
            } else {
                ChinesePinyinSeedDictionary.dict
            }
            val engine = DictionaryCandidateEngine(dict)
            cachedEngine = engine
            cachedKey = key
            return engine
        }
    }
}

File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/Lang/ChineseLanguagePack.swift
import Foundation

/// Chinese (中文) pinyin pack — type pinyin on the standard QWERTY;
/// `PinyinComposer` shows the pinyin live and `DictionaryCandidateEngine` offers
/// Hanzi candidates. Same shape as Japanese, only the dictionary differs.
/// Mirrors the Kotlin ChineseLanguagePack.
public struct ChineseLanguagePack: LanguagePack {
    public let id = "zh"
    public let displayName = "中文"
    public let locale = Locale(identifier: "zh")
    public let longPressAccents: [Character: [Character]] = [:]

    /// App Group container — set by KeyboardViewController, same as JP/EN.
    public static var appGroupContainer: URL?
    /// Matches the backend manifest's pack URL prefix (zh pack ships later).
    private static let packIdPrefix = "draftright-ime-zh"
    /// Cached engine — parse the pinyin pack once per session, not per keystroke.
    /// Cached engine + the pack identity it was built from. Parsing the pack is
    /// expensive, so keep the engine across keystrokes — but key the cache on
    /// the resolved pack URL (its filename encodes the version) so installing a
    /// newer pack mid-session rebuilds instead of serving the stale seed/dict
    /// (issue #10). Per-keystroke cost is one small directory scan.
    private static var cachedEngine: CandidateEngine?
    private static var cachedKey: String?

    public init() {}

    public func makeComposer() -> Composer? { PinyinComposer() }

    public func makeCandidateEngine() -> CandidateEngine? {
        let key = DictPackResolver.resolvedPackURL(
            appGroupContainer: Self.appGroupContainer,
            packIdPrefix: Self.packIdPrefix
        )?.path ?? "seed"
        if key == Self.cachedKey, let cached = Self.cachedEngine { return cached }
        let dict = DictPackResolver.loadOrFallback(
            appGroupContainer: Self.appGroupContainer,
            packIdPrefix: Self.packIdPrefix,
            fallback: { ChinesePinyinSeedDictionary.dict }
        )
        let engine = DictionaryCandidateEngine(dictionary: dict)
        Self.cachedEngine = engine
        Self.cachedKey = key
        return engine
    }

    public var alphaRows: [[KeyDef]] { QwertyLayout.alphaRows }
    public var symbols1Rows: [[KeyDef]] { QwertyLayout.symbols1Rows }
    public var symbols2Rows: [[KeyDef]] { QwertyLayout.symbols2Rows }
}
